build-tools/packages/build-tools/src/type-test-generator/main.ts

```diff
@@ -0,0 +1,177 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
+import path from "node:path";
+import { Project, SourceFile } from "ts-morph";
+import { BrokenCompatTypes } from "../common/fluidRepo";
+import { PackageJson } from "../common/npmPackage";
+import { buildTestCase, TestCaseTypeData } from "../typeValidator/testGeneration";
+import { getFullTypeName, getNodeTypeData, TypeData } from "../typeValidator/typeData";
+
+// Do not check that file exists before opening:
+// Doing so is a time of use vs time of check issue so opening the file could fail anyway.
+// Do not catch error from opening file since the default behavior is fine (exits process with error showing useful message)
+const packageFile = readFileSync("package.json");
+
+const packageObject: PackageJson = JSON.parse(packageFile.toString());
+
+const testPath = `./src/test/types`;
+// remove scope if it exists
+const unscopedName = path.basename(packageObject.name);
+
+const fileBaseName = unscopedName
+	.split("-")
+	.map((p) => p[0].toUpperCase() + p.substring(1))
+	.join("");
+const filePath = `${testPath}/validate${fileBaseName}Previous.generated.ts`;
+
+if (packageObject.typeValidation?.disabled) {
+	console.log("skipping type test generation because they are disabled in package.json");
+	// force means to ignore the error if the file does not exist.
+	rmSync(filePath, { force: true });
+	process.exit(0);
+}
+
+const previousPackageName = `${packageObject.name}-previous`;
+
+{
+	// Information about the previous package from the package.json is not needed,
+	// but error if its missing since its nice to septate errors for the dep missing here vs not installed.
+	const previousDep = packageObject?.devDependencies?.[previousPackageName];
+	if (typeof previousDep !== "string") {
+		throw new Error(`Did not find devDependency ${previousPackageName} in package.json`);
+	}
+}
```

Comment:
this is a common pattern when you have a local that you don't need to be in-scope below. The existing comment notes that this is not needed below, and putting it in a nested scope uses the compile to enforce this (and thus helps the reader trust the comment) and avoid accidental changes which would make the comment untrue.

I don't think describing this pattern every time we use it would be a new improvement for readability, but if you have a suggestion to how we can help clarify things here concisely I'm open to suggestions.

Reply:
This is the first time I've seen this done in JS, hence my question. If we commonly do this that's fine.